### ISSUE TITLE
Docker process doesn't free up memory - macOS, Apple Silicon, Virtualization.framework

### ISSUE THREAD

**Issue by z1r0-** [docker/for-mac#6120]
With Virtualization.framework enabled the Docker process (not Virtual Machine Service) grows in memory-usage and doesn't free it up properly. After some testing I noticed, that it seems to be related to transfers between host and docker-container.  
Easiest example: copy a 1GB file using "docker cp" and the memory-usage will grow by 1GB. But happens with other stuff too (like importing a db with local client over a host-bind port).

Example:
```
dd if=/dev/zero of=output.dat  bs=1m  count=1024
docker run -d --rm --name mem_test alpine tail -f /dev/null
docker cp output.dat mem_test:/tmp/
# Watch Memory grow in Activity Monitor
```

I've seen the discussion about "Memory" and "Real Mem" in Activity Monitor and first thought "okay, Real Mem returns to normal after some time so it will be fine" but then I noticed, that MacOS starts swapping when "Memory" value gets to big even "Real Mem" is just some MB.

<img width="688" alt="image" src="https://user-images.githubusercontent.com/6104256/147954171-31543408-433b-4ccb-b22d-b00ae9b482a3.png">

<img width="595" alt="image" src="https://user-images.githubusercontent.com/6104256/147954209-9b305792-8aaf-42ac-9ee8-f4ea6366c693.png">

After restarting Docker (from docker context menu) everything goes back to normal again.. So I'm currently restarting Docker every time memory usage gets to huge.. I've usually 20+GB of free memory and no swap usage if I don't forget to restart Docker, so it's deff it.

I'm using the most recent version: Docker Desktop 4.3.2 (72729) and MacOS 12.2 Beta (but I had same behaviour with 12.0, 12.1).

**Comment by djs55** [comment]
@z1r0- thanks for the report. Could you upload a set of diagnostics and a matching screenshot? I'd like to confirm exactly which process is leaking (and view the command-line etc)

**Comment by z1r0-** [comment]
@djs55 Sure. Diagnostics ID: 72F4E822-C849-4117-981D-06B38A63C1BA/20220106145448

![image](https://user-images.githubusercontent.com/6104256/148402612-3083afb8-4f6e-4c57-a919-2e615c09ea81.png)

Used docker cp to let it grow (like in the example above) and stopped/removed the container.

**Comment by bjonnh-work** [comment]
I can confirm that, let me know if you need any additional info/testing/logs.

**Comment by joaopbnogueira** [comment]
Pitching in as I have to restart docker multiple times a day to work around this issues

I'm currently on the experimental build 4.6.0 (75045) running 12.2.1 on an M1 Pro with VirtioFS enabled, but the behavior is the same on the GA 4.5.0 version.

Diagnostics ID AF628F15-7B04-4806-BEE7-33B5827B55A8/20220309231313

<img width="769" alt="Screenshot 2022-03-09 at 23 21 09" src="https://user-images.githubusercontent.com/1637158/157555761-3c5090ab-1d67-4d71-95aa-65ada02d8b52.png">

**Comment by bjonnh-work** [comment]
75324 has been well behaved for me for the last few days (M1 and VirtIO FS enabled)

**Comment by z1r0-** [comment]
uh. completely missed, that there's a 75324 build. just downloaded it and made the docker cp test and the issue sadly is still there. 

> 75324 has been well behaved for me for the last few days (M1 and VirtIO FS enabled)

It completely depends on what you are doing. On days where I only need to code it works without any issues and I can let it run for days. But when I need to import db-dumps > 15GB it's a huuuge issue. Started to copy the gziped dumps to the db container and then import them instead of using local psql client to reduce the issue but it still remains a problem.

Something else: I'm also using mutagen for better io-performance and somehow it seems not to be affected by this issue. It's only "docker cp" and "shared ports" that I identified so far.

**Comment by tysliu** [comment]
M1 and VirtIO FS enabled. Docker version 4.6.0 (75818). A high disk IO app runs for 5 minutes docker process ends up using 6 GB of ram.

**Comment by joaopbnogueira** [comment]
Checking in as this issue is still present on 4.6.1.

Docker process using 103GB (and swapping a lot of course).

Diagnostics ID: AF628F15-7B04-4806-BEE7-33B5827B55A8/20220331104836

**Comment by marcoscale98** [comment]
Alsa without virtualization.framework!

![Schermata 2022-04-05 alle 22 03 55](https://user-images.githubusercontent.com/43850400/161840831-ae3d894c-7e9d-47de-b5a6-4707b873c5d1.png)
![Schermata 2022-04-05 alle 22 03 21](https://user-images.githubusercontent.com/43850400/161840837-2cd98f50-4173-4021-b1a4-70b3f6892d6a.png)
![Schermata 2022-04-05 alle 22 03 48](https://user-images.githubusercontent.com/43850400/161840842-07a7c910-6759-4865-b715-23dd582c3bf8.png)

**Comment by yunussandikci** [comment]
I am having the same issue with Docker Desktop 4.7.0 (77141) 
Diagnostic ID: 79C10953-31E7-4308-9955-4FFEF6A3129F/20220418100515

![Screen Shot 2022-04-18 at 13 06 58](https://user-images.githubusercontent.com/23269628/163793263-a0126269-9fe2-4594-aca5-cf3bd83d1a55.png)

**Comment by beatslikeahelix** [comment]
What actually is Virtual Machine Service?

**Comment by balabis** [comment]
Having exactly the same issues on my m1, sadly.

**Comment by joaopbnogueira** [comment]
Yup, stil present on 4.8.1

**Comment by benwixen** [comment]
This happens for me as well on Docker Desktop 4.8.2 on an M1 Mac running macOS 12.4.

If I enable Virtualization.framework and `docker cp` or wget a 100mb file into the container, the Docker process will also grow its memory footprint by 100mb.

Seems to be related to network traffic - I was able to move the same file into the container via a bind mount without the same thing happening.

**Comment by jonwest** [comment]
Still happening on `4.9.1` as well on an M1 Mac running macOS 12.4.  

Any sort of network traffic seems to grow the memory usage over time.  Running things like `cp` or `dd` do not grow the memory.

**Comment by AiAe** [comment]
Docker 4.10.1
M1 Pro with 32GB
MacOS 12.4 
Docker itself does not use a lot of memory, but the swap goes up to 60gb and it goes down after Docker restart.
CPU usage is also really high when this happens.
Tested with and without the experimental features.

**Comment by benwixen** [comment]
@djs55 Is there anything more needed for this issue to be acknowledged? Let us know if there's anything we can do - I have several people on my team struggling with this issue.

**Comment by mkellerman** [comment]
Docker 4.10.1 (82475)
Engine: 20.10.17

Mac Studio M1 Ultra - 128 GB

<img width="873" alt="image" src="https://user-images.githubusercontent.com/9411143/180041513-764a7bdd-9472-4c12-bf87-c0d0819b9897.png">

<img width="1154" alt="image" src="https://user-images.githubusercontent.com/9411143/180042418-4c9b83f0-5550-402d-baf7-8786d7f700e3.png">

I used to run PostgreSQL in docker as well, but docker would eventually crash. I now run PostgreSQL directly on the host.

**Comment by dyanakiev** [comment]
Docker 4.10.1
M1 Air with 16GB
MacOS 12.4

Got my ram filled up my laptop crashed and broken folder icon appeared... lost all my projects and documents.. worst experience ever!
 
Please fix ASAP.

**Comment by RemiKalbe** [comment]
Having the same issue, (64GB m1 max).
In less than a minute it uses all my memory and my computer freezes, the only thing I can do is a hard reset.
```
Client:
 Cloud integration: v1.0.24
 Version:           20.10.17
 API version:       1.41
 Go version:        go1.17.11
 Git commit:        100c701
 Built:             Mon Jun  6 23:04:45 2022
 OS/Arch:           darwin/arm64
 Context:           default
 Experimental:      true

Server: Docker Desktop 4.10.1 (82475)
 Engine:
  Version:          20.10.17
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.17.11
  Git commit:       a89b842
  Built:            Mon Jun  6 23:01:01 2022
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.6
  GitCommit:        10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
 runc:
  Version:          1.1.2
  GitCommit:        v1.1.2-0-ga916309
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Comment by artemijan** [comment]
Seems like disabling Virtualisation framework resolves the issue.

**Comment by artemijan** [comment]
Don't know why I got 5 dislikes, but to resolve the issue right now you have to disable Virtualisation or restart docker once notice swap usage))
<img width="938" alt="image" src="https://user-images.githubusercontent.com/7948990/181617486-a3caa6bc-e17b-4460-99ab-a3cdab120b21.png">
Without virtualisation framework qemu process always take 2,5GB, and there is no grow in memory usage

**Comment by havekes** [comment]
> Don't know why I got 5 dislikes, but to resolve the issue right now you have to disable Virtualisation or restart docker once notice swap usage))
> <img alt="image" width="938" src="https://user-images.githubusercontent.com/7948990/181617486-a3caa6bc-e17b-4460-99ab-a3cdab120b21.png">
> Without virtualisation framework qemu process always take 2,5GB, and there is no grow in memory usage

On Apple Silicon, the read performance on mounted volumes is absolutely trash without VirtioFS accelerated directory sharing, for which you need to enable the new Virtualization framework.

**Comment by artemijan** [comment]
I haven’t noticed it, I’m just running redis and postges for my app, and it works pretty fine. But I’ll doublecheck it. Thanks

**Comment by pythoninthegrass** [comment]
Seeing high RAM usage on an Intel Mac running macOS Monterey 12.3.1 (21E258), [this thread](https://forums.docker.com/t/macos-high-com-docker-hyperkit-cpu-usage-when-no-containers-running/115705/12) helped with debugging VM processes:
> I can only guess that the virtual machine itself stuck in a state it couldn’t properly start but couldn’t stop either.
> Next time it happens try this:
> ```bash
> cd ~/Library/Containers/com.docker.docker/Data/vms/0
> nc -U console.sock
> ```
> 
>Then press ENTER to get the console prompt.
>Press CTRL+u to delete the special charcters appeared in the console like: `^[[51;19R[32`
> 
> Run top tor other monitoring command to find what process uses the CPU.
> The console may continuously shows you error messages or warnings. If there aren’t too many and not relevant, press ENTER and ignore it. If there are too many you probably should deal with it.
> You can also try the debug shell:
> ```bash
> cd ~/Library/Containers/com.docker.docker/Data
> nc -U debug-shell.sock
> ```
> You can use it similarly to the console except it won’t show you error messages over the prompt. However, some commands works differently here like runc list but you can see the same processes. I think it just sees the resources and processes in the VM.

What ultimately helped is dropping Docker resources:
* **CPUs**: 4 > 2
* **Memory**: 4.00 GB > 2.00 GB
* **Swap**: 1 GB > 512 MB

![Screen Shot 2022-08-02 at 3 13 21 PM](https://user-images.githubusercontent.com/4097471/182465001-cdc7858a-7718-4836-877d-b28316443de9.png)

After it restarted, that freed up swap by ~10GB. Could probably also do `sudo purge`, but try to avoid that. Virtual Machine Service is now consistently at 2.00 GB after stopping containers. Hyperkit is being used by `minikube`, so irrelevant here.

![Screen Shot 2022-08-02 at 3 13 53 PM](https://user-images.githubusercontent.com/4097471/182465003-3626a21b-8e32-4c8c-ae76-465deed44528.png)

Disabled gRPC FUSE, enabled VirtioFS, telemetry was already disabled. Not sure if the former two had much of an effect on CPU/RAM usage vs. IO which wasn't tested.

Would prefer to not hamstring containers on a fairly specc'd out Mac mini, but satisfied for now.

**Comment by NeoRoth** [comment]
Same issue on my M1 max.

![at start](https://user-images.githubusercontent.com/18250033/184157498-d7104539-ed3d-45fc-9514-5be4cf756443.png)
few moment later
![few moment later](https://user-images.githubusercontent.com/18250033/184157545-947751f3-aa7a-4e41-a34b-94e9aa0a1d98.png)

All the data downloaded by the container is counted in the memory, remains counted in the real memory then is finally compressed by the operating system which tries to survive. It all ends up in the swap until the host machine crashes.

**Comment by dvcrn** [comment]
Copied from my other issue:

When the container writes into big file on a mounted path, memory usage of Docker keeps going up, and overshoots the limit set in 'Resources', effectively taking the entire Mac down until it's force closed

I've set 8GB memory limit in Resources. Container is mounting a few things:

```
    volumes:
      - ~/docker/config:/config:cached
      - /Volumes/dockerroot:/data:delegated
```

The container is trying to write a 12gb file (downloaded / streamed) over time into `/data`. As the file gets written, memory on the host shoots up drastically. To my naive eyes it looks like it's trying to keep the entire content in memory, instead of periodically flushing the state to disk. 

Memory on the Docker process itself is going up, not on the virtual machine service 

<img width="401" alt="Screen Shot 2022-10-24 at 20 42 21" src="https://user-images.githubusercontent.com/688326/197517932-e96adc5f-493d-4642-8208-9f1c53ebac6c.png">


`docker stats` is reporting nothing strange besides that NET I/O is very similar to memory used by the process 
```
CONTAINER ID   NAME                  CPU %     MEM USAGE / LIMIT    MEM %     NET I/O           BLOCK I/O        PIDS
98b265435c76   mycontainer          8.79%     14.95MiB / 7.75GiB   0.19%     2.61GB / 75.8MB   0B / 336kB       14
```

Basically as previous posters said, this just climbs and climbs until the host starts randomly killing things. Glad I was able to finally see what's going on as this usually happened on my mac mini while I was sleeping 😅

**Comment by z1r0-** [comment]
I updated to macOS Ventura yesterday and retested it (with docker cp like described in my initial post) and the issue seems to be gone? Can anyone else confirm this?

**Comment by rubnogueira** [comment]
![image](https://user-images.githubusercontent.com/40404708/198012823-66a9d2d9-9d9d-4dca-ad7e-d150490971de.png)

It looks its fixed!

**Comment by rubnogueira** [comment]
I really think it's fixed.

![image](https://user-images.githubusercontent.com/40404708/198013370-047cff49-c6f2-48a9-8a4d-eb10da329eee.png)

**Comment by stefanomaxenti** [comment]
Problem still present in 4.13.1 and Mac M1.

**Comment by rubnogueira** [comment]
> I really think it's fixed.
> 
> ![image](https://user-images.githubusercontent.com/40404708/198013370-047cff49-c6f2-48a9-8a4d-eb10da329eee.png)

Unfortunately, I was wrong. The problem is still present :\

**Comment by dvcrn** [comment]
Not fixed here either, on Ventura 0.1. Still need to manually clean up Docker once in a while on my server because the memory usage is becoming too huge. So don't upgrade to Ventura hoping it'll make things better 😇

**Comment by pransh15** [comment]
Yeah - can confirm what @dvcrn said. 
![image](https://user-images.githubusercontent.com/18158596/202579828-a4fb2f26-191d-4632-94be-c9ea212d7ad1.png)

**Comment by guillermoeuillades** [comment]
I have same issue in a intel based macbook pro. after stop all contairnsers it's reaming using full memory. 

<img width="724" alt="Captura de Pantalla 2022-11-19 a la(s) 16 43 49" src="https://user-images.githubusercontent.com/52510298/202868648-8c66e0f6-2faa-4383-bca2-70bd924bb412.png">

**Comment by laviua** [comment]
Ventura 13.0 (22A380)
M1 mac mini

The same problem

**Comment by davidaberg** [comment]
This will not solve the real problem but I have found a workaround that helps me get by without the Docker app memory building until the Docker daemon crashes. For me that takes approximately 24 hours before the daemon becomes unresponsive.
So I have added this commands to my crontab (-e) that stops Docker 04:00 and then starts it 04:02 every night.
0 4 * * * killall Docker
2 4 * * * open -a Docker
(Don't chain the commands with a sleep between, didn't work for me).
This basically restarts Docker every night automatically to free up the unreleased Swap memory from Docker.app.
Of course far from optimal for Production systems, but hey, better than crashing.
You might want to shorten the time between restarts depending on how long your system is alive before crashing.
I run Docker 4.14.1 that comes with Virtualization.framework enabled by default. I have disabled VirtioFS for now.
Seems to make Docker build up the Swap and crash even faster.

**Comment by tonilampela** [comment]
My M1 mac mini was acting wonky today, and during investigation noticed Docker consuming 26GB of RAM with bunch of swapping happening. I just started to use Docker yesterday on this machine. Running macOS Ventura 13.0.1.

Restarting helped for now, but clearly this is a serious issue when running Docker on Mac.

**Comment by dstoyanoff** [comment]
Jumping in with a similar problem. I have just noticed the process `Virtual Machine Service` always taking 8.01GB RAM. Restarting docker kills the process, but when Docker starts, it goes to 8.01GB. This is before I even try to run a container.

**Comment by joaopbnogueira** [comment]
@dstoyanoff I think that is a different issue. Virtualization Framework allocates the full amount of memory that you specify for docker at start, but this issue is about the actual Docker process having a memory leak.

**Comment by tysliu** [comment]
4.15.0 seems to have resolved it! Virtualization Framework is enabled by default. But if I enable VertioFS for file sharing Docker fails to start. Suppose you are stuck and unable to change the filesharing implementation. Try...
```
vi ~/Library/Group\ Containers/group.com.docker/settings.json
"useVirtualizationFrameworkVirtioFS": false,
```
M1 MacMini MacOS 12.6

**Comment by tysliu** [comment]
Apologies false report. After leaving it for an hour it's now taking up 46GB of memory.

**Comment by ahmedsayedabdelsalam** [comment]
M1 MacMini MacOS 13.0.1
I ended up with ~2TB SSD Data Usage cuz of swap 😢

**Comment by euroegyptn** [comment]
I'm also seeing this issue on my Mac mini - Intel

macOS 13.1
Docker Desktop version v4.15.0

**Comment by wysiecki** [comment]
same here, MBP M1 16GB 
swap used 27GB, my container is generating report files, with 13MB all of them, so not that much

macOS 13.1
Docker Desktop version v4.15.0

:-(

**Comment by z1r0-** [comment]
@eshirvana That's not an issue, that's how it's intended to work. Virtual Machine Service allocates the full amount of memory that you set in your docker settings. But that's normal.

The problem that is talked about here is about the "Docker" process that grew over time. 

Still wondering what the issue for it is and why it stopped for me after updating to Ventura but some still experiencing it.

Just a wild guess: Can it somehow be related to xcode/xcode-tools version installed? At the time Ventura was released the new xcode version was still beta. I had installed it, because it was needed for homebrew to work. My thought was that maybe a newer version of xcode updates stuff for the Virtualization.Framework or is it bundled with the app? I'm not too much into macos development so I don't really now how it works..

**Comment by dabvid** [comment]
Did anyone try to resolve this by emulating x86 with Rosetta that recently was introduced?
Not sure if it's worth the trouble but since Apple Silicon seems to be a factor I guess it could be worth trying out.

**Comment by Kectai** [comment]
Docker Desktop 4.16.0 (95345)
<img width="523" alt="image" src="https://user-images.githubusercontent.com/46049250/212468343-fede0d26-6055-4789-8805-43f74e413d5f.png">
<img width="717" alt="image" src="https://user-images.githubusercontent.com/46049250/212468385-cf1b3869-175f-46bf-9519-6bdbfcb52fc3.png">

**Comment by wysiecki** [comment]
NICE ! Working now as expected, thank YOU !
Docker Desktop 4.16.1
MBP M1 16GB

**Comment by oloushkin-ah** [comment]
i can't confirm that.
the same problem in both framework options. Virtualization framework enabled/disabled

**Comment by EliranTurgeman** [comment]
same issue here. on intel chip 
Does someone knows if there is a workaround

**Comment by dabvid** [comment]
Nice! Finally no memory leakage, at least not for me on M1 Mac Mini. Great work with the latest release 🙏

**Comment by lucax88x** [comment]
Had to switch to podman, and it's however in general consuming less memory.

**Comment by laviua** [comment]
> Nice! Finally no memory leakage, at least not for me on M1 Mac Mini. Great work with the latest release 🙏

m1 mac mini, still the same memory consumption

**Comment by dabvid** [comment]
> > Nice! Finally no memory leakage, at least not for me on M1 Mac Mini. Great work with the latest release 🙏
> 
> m1 mac mini, still the same memory consumption

Sorry to hear. Are you saying that the memory Docker uses on your machine keeps growing until your computer crashes? In that case, what file system are you using? Do you have Virtualization framework enabled?

**Comment by laviua** [comment]
thanks for forcing me to show the proofs :)
I've tried to reproduce the leaking by the first comment and it is not leaking anymore, set 4 GB allocated memory for the docker process and it's not growing.

> > > Nice! Finally no memory leakage, at least not for me on M1 Mac Mini. Great work with the latest release 🙏
> > 
> > 
> > m1 mac mini, still the same memory consumption
> 
> Sorry to hear. Are you saying that the memory Docker uses on your machine keeps growing until your computer crashes? In that case, what file system are you using? Do you have Virtualization framework enabled?

**Comment by byre54** [comment]
M1 Pro 16GB RAM
<img width="1041" alt="image" src="https://user-images.githubusercontent.com/5574999/212970789-3b90c9ce-ed9c-4022-862e-839c5627f791.png">

6GB spent even I restarted the docker after update. 
energy consumption also impacted. I prefer using docker on linux than OSX atm.

**Comment by rubnogueira** [comment]
@byre54 have a look at this comment https://github.com/docker/for-mac/issues/6120#issuecomment-1373353088

**Comment by aalimovs** [comment]
> M1 Pro 16GB RAM <img alt="image" width="1041" src="https://user-images.githubusercontent.com/5574999/212970789-3b90c9ce-ed9c-4022-862e-839c5627f791.png">
> 
> 6GB spent even I restarted the docker after update. energy consumption also impacted. I prefer using docker on linux than OSX atm.

You need to right on column names and click "Real Memory". The 6 GB you see is the memory reserved for it, the actual memory usage will be lower.

**Comment by aalimovs** [comment]
> M1 Pro 16GB RAM <img alt="image" width="1041" src="https://user-images.githubusercontent.com/5574999/212970789-3b90c9ce-ed9c-4022-862e-839c5627f791.png">
> 
> 6GB spent even I restarted the docker after update. energy consumption also impacted. I prefer using docker on linux than OSX atm.

@byre54 You need to right on column names and click to show "Real Memory". The 6 GB you see is the memory reserved for it, the actual memory usage will be lower.

**Comment by jay-hub** [comment]
The default value when installing docker seems to be somewhere around 4.5GB. It's possible to change this using docker according to the need. 

<img width="835" alt="Screenshot 2023-03-08 at 8 32 13 PM" src="https://user-images.githubusercontent.com/705023/223892203-74b72d3f-c69b-40c9-8f55-af8d232adfd8.png">

**Comment by Ao-Last** [comment]
Same issue here. Does it mean that the virtual machine hangs once I start docker desktop except I manually quite the VM service?

**Comment by GOWxx** [comment]
same issue.

macos 13.1 (22C65)

docker desktop Version 4.17.0 (99724)
Engine: 20.10.23
Compose: v2.15.1
Credential Helper: v0.7.0
Kubernetes: v1.25.4

**Comment by radio1988** [comment]
Thank you jay-hub, restricting the memory allocated to Docker helped me a lot

**Comment by aabouelleil-seniorlink** [comment]
I have also seen this issue. Let me know if you need any additional data.

**Comment by tbrandt-circle** [comment]
Still having this issue.

macos 13.3

docker desktop Version 4.17.0 (99724)
Engine: 20.10.23
Compose: v2.15.1
Credential Helper: v0.7.0
Kubernetes: v1.25.4

**Comment by realhuseyn** [comment]
I also have this issue.
MacOS 13.3
Docker Desktop version v4.12.0
Docker Engine v20.10.17
Virtual Machine Service is consuming 7.91 GB. But I noticed 13 GB consumption previously (before restarting Docker Desktop). 

<img width="839" alt="image" src="https://user-images.githubusercontent.com/31611637/229309178-98838483-a8e4-4958-95cb-29590db2ceb3.png">

Also, I checked "Resources" settings in the Docker settings. I saw Memory: 7.90 GB. But I never configured this setting, it is by default settings. Should I decrease this setting? Will Docker work normally after decreased memory settings?

<img width="982" alt="image" src="https://user-images.githubusercontent.com/31611637/229309411-03a92147-c0fd-4e1a-863d-a07132f491a4.png">

**Comment by jay-hub** [comment]
> I also have this issue. MacOS 13.3 Docker Desktop version v4.12.0 Docker Engine v20.10.17 Virtual Machine Service is consuming 7.91 GB. But I noticed 13 GB consumption previously (before restarting Docker Desktop).
> 
> <img alt="image" width="839" src="https://user-images.githubusercontent.com/31611637/229309178-98838483-a8e4-4958-95cb-29590db2ceb3.png">
> 
> Also, I checked "Resources" settings in the Docker settings. I saw Memory: 7.90 GB. But I never configured this setting, it is by default settings. Should I decrease this setting? Will Docker work normally after decreased memory settings?
> 
> <img alt="image" width="982" src="https://user-images.githubusercontent.com/31611637/229309411-03a92147-c0fd-4e1a-863d-a07132f491a4.png">

I believe the required memory depends on your requirement and the containers you're running on at a time. In my case, it's AWS Lambdas (using AWS SAM CLI) and a PostgreSQL container (which in combination doesn't require many resources. Feel free to play around to find the sweet spot for the requirement.

**Comment by marcelosalloum** [comment]
I've found the issue and Docker doesn't seem to be the sole culprit in my case.

**TL;DR**: I reduced the memory available to Docker from 8GB (50% of my machine memory) to 4GB and it solved the issue! 🎉 🚀 

I was giving Docker a max memory of 8GB, but I didn't realize that Docker would take that amount of memory for itself even when that amount wasn't needed. This made my computer use close to 100% of memory, which froze the `docker-compose up` command I was trying to execute.

When I downgraded Docker's memory to 4GB, my computer went back to run with a healthier 60%-80% memory usage, which unblocked the processes I was trying to run.

**Comment by muescha** [comment]
> - Settings / Resources / Advanced:
>	- CPU: 5
>	- Memory 8GB
>	- Swap 1GB
>	- Virtual disk limit 64

did you mean this settings?

what here your settings?

**Comment by marcelosalloum** [comment]
> > * Settings / Resources / Advanced:
> >   
> >   * CPU: 5
> >   * Memory 8GB
> >   * Swap 1GB
> >   * Virtual disk limit 64
> 
> did you mean this settings?
> 
> what here your settings?

@muescha: Exactly, I changed my memory settings at `Settings / Resources / Advanced` from `Memory: 8GB` to `Memory: 4GB` and now my docker-compose jobs run smoothly.

**Comment by muescha** [comment]
Your other values are the same?

**Comment by marcelosalloum** [comment]
@muescha: here are all the values:

<img width="815" alt="Screenshot 2023-04-07 at 2 21 06 PM" src="https://user-images.githubusercontent.com/1952597/230680465-aa673277-581d-48b9-9925-9b3dda6fede3.png">

**Comment by JoeYang1412** [comment]
Sorry, I recently started using Docker, and my question is, is the memory usage of this virtual machine service program that I see in the Mac's Activity Monitor the amount I set in Docker? Why is it using up all the memory or even more?

<img width="1000" alt="截圖 2023-05-02 下午4 32 50" src="https://user-images.githubusercontent.com/94301119/235619799-fcbf087b-22b5-4384-b22a-8f221f925cd5.png">
<img width="1000" alt="截圖 2023-05-02 下午4 33 05" src="https://user-images.githubusercontent.com/94301119/235619803-06b49393-7729-4bd6-8d49-0b45a45336fa.png">

**Comment by joaopbnogueira** [comment]
@JoeYang1412 when using Virtualization Framework, yes.

But i just want to point out to everyone that this issue is about the Docker process itself not releasing memory and keeping growing and has nothing to do with the memory that needs to be allocated to actually run the containers which seems to be allocated up front with Virtualization Framework.

FYI this seems to have improved for me in more recent builds and I haven't seen the Docker process eating up 30GB+ of RAM lately.

**Comment by joaopbnogueira** [comment]
@JoeYang1412 exactly, that looks fine 👍

**Comment by JoeYang1412** [comment]
Are you referring to this? But it seems to look normal, so my Docker is currently functioning properly? There are no issues, right?
<img width="1000" alt="截圖 2023-05-02 下午8 45 00" src="https://user-images.githubusercontent.com/94301119/235670197-8e128693-e199-486a-b404-efa23c10c188.png">

**Comment by joaopbnogueira** [comment]
Exactly, that looks good!

**Comment by JoeYang1412** [comment]
@joaopbnogueira 
Okay, thank you for your help. Have a great day!

**Comment by joaopbnogueira** [comment]
Apparently, I was wrong. Even though on recent versions the Docker process does not balloon in terms of memory allocation, the `com.docker.backend` does, so it seems that the memory growth issue was simply moved to another process.

Diagnostics ID: AF628F15-7B04-4806-BEE7-33B5827B55A8/20230503100400


<img width="779" alt="Screenshot 2023-05-03 at 11 03 31" src="https://user-images.githubusercontent.com/1637158/235887666-2c97ca9a-724c-453f-8a34-8062c85d977e.png">

<img width="750" alt="Screenshot 2023-05-03 at 11 03 54" src="https://user-images.githubusercontent.com/1637158/235887682-1e769e57-2956-4896-a5bc-8e3b9d6f15af.png">

**Comment by Oluwatobilobaoke** [comment]
Exactly, temporarily solves it, Docker as a firm needs to find a fix to this issue

**Comment by unieagle** [comment]
The issue still here.
Limit mem cap won't help if you want to start a container, but docker mem usage is close or exceed the cap.
In my case, the pure docker takes up 12GB memory after closed all running containers. If I set the limit to 12GB, then, there will be some error when trying to start container.
Docker need to release the mem. Sadly, the currently only solution is to restart docker manually each time it reaches the cap.

**Comment by mac-repo-001** [comment]
Issue still present as of today on latest macOS 13.4.1 on Intel iMac 27" 2019

**Comment by developedsoftware** [comment]
Still happening for me too on latest version of docker and an m2 mac

What's the solution ?

**Comment by doulighan** [comment]
Still happening. I open docker desktop and within 2 hours my code editors grind to a halt. Force quitting every docker process immediately fixes it.

**Comment by tuan-phan-spenmo** [comment]
are there any solutions for that?

**Comment by ASRodrigo1** [comment]
Same issue is still happening, also in Linux and Windows.

**Comment by SmartArray** [comment]
<img width="865" alt="Screenshot 2023-07-21 at 09 46 09" src="https://github.com/docker/for-mac/assets/934592/e2699a33-3faf-4914-b73e-f8002b3212ab">
Guys, this is getting out of hand 🥲

**Comment by aktienautobahn** [comment]
i am using Docker Desktop 4.21.1 (114176) on macos Ventura 13.4.1 (c) M1 Pro. Docker (Desktop?) let a user statically allocate memory, whether 2 GB, 4 GB or more. process memory usage on host grow almost immediately to the max memory allocation(defined in Docker Desktop Setting in the Resources page) with no containers running. That means, there is not possible to let Docker manage containers resources usage dynamically. Really bad user experience. Any Linux Users with better user experience? Does Docker on Linux dynamically manage memory usage? I am also trying to use Docker Desktop alternative, but it is still in beta and the developer is not well-known: https://orbstack.dev

**Comment by ibash** [comment]
Docker 4.22.0 with resource saver will shutdown virtual machine service when not in use: https://www.docker.com/blog/docker-desktop-4-22/

**Comment by ships** [comment]
Can confirm that before 4.22 my system was always pegged to maximum allowed memory in docker desktop settings (32Gb in my case) and I would normally kill docker and restart  it to clear. But after upgrading to 4.22.0 and without killing the VM Service, it dropped from 32 to 3Gb.

**Comment by hibiki-kato** [comment]
I have the same(?) issue.
Docker 4.24.0
MacOS Sonoma
By updating docker, memory usage of docker displayed properly compared to the former situation: Activity monitor always showing docker using the maximum size of allocated memory, but the MEMORY PRESSURE graph shows proper memory usage.
Initially, memory usage by "Virtual Machine Service for Docker" was small, but after running C++ codes that utilize a large amount of memory, it became large and seemed not released as shown below.
![Screenshot 2023-10-02 at 17 00 56](https://github.com/docker/for-mac/assets/125548492/438bdcb0-c754-421e-b806-01db2363109f)
I suspected that a memory leak happened in my code at first, but the actual memory size seemed small according to the "docker stats".
In the end, I'm still unsure whether the memory is actually being used or not.

**Comment by jumarko** [comment]
From https://news.ycombinator.com/item?id=34719785
> **It's page cache**. A similar thing happens with wsl2 on Windows[1]. The issue is the linux kernel running in the lima-vm thinks all the ram you give the VM is free to cache files on, it doesn't know there is a host OS that could also make use of that ram. So after loading up a ton of containers where each has its own OS dependencies that ends up being a lot of files the kernel keeps cached so long as nothing else needs the memory. There isn't really a fix besides right sizing the amount of ram you give the VM. Or I guess you could disable file caching but I wouldn't recommend it[2].

**Comment by eiko4** [comment]
I'm still having problems with RAM.

This is a bug in MacOS with Dynamic memory allocation.
True dynamic memory allocation would allow Docker Desktop to release memory back to macOS when it's no longer needed, reducing memory usage and removing the need to adjust the limit.

Unfortunately, dynamic memory allocation is not production-ready due to a still-unfixed macOS bug (FB11996689). 

It would be great if someone would open more tickets with Apple. Because for some reason the current ticket is not a priority for them.

**Comment by krzysztoflupa** [comment]
The problem still exist and it's enormous in some cases, @eiko4 how can we track the progress of FB11996689 and put some pressure on apple??

**Comment by eiko4** [comment]
@krzysztoflupa This is a private issue in the bug tracker - we cannot track the status. However, the Docker for Mac team can open a new issue in the apple bug tracker to track this.

Guys, I hope that you write to Apple about this. Otherwise, we won't see a fix soon, if at all.

If you do this, please let us know here.

**Comment by philipkocanda** [comment]
I think I've managed to confirm the network traffic hypothesis:

1. Expose the Docker socket on `localhost:2376` so we can run apachebench against it:

```sh
socat TCP-LISTEN:2376,fork UNIX-CONNECT:/var/run/docker.sock
```

2. Open up another terminal, and run apachebench to simulate a lot of network traffic in a short amount of time:

```sh
ab -r -n 10000000 -c 4 -v 1 http://127.0.0.1:2376/containers/json
```

3. Notice the memory consumption increase to almost 1GB after only four minutes of requests:

**Initial memory footprint:**

<img width="1072" alt="Screenshot 2024-01-08 at 16 28 58" src="https://github.com/docker/for-mac/assets/1023577/13e169b9-922f-4674-a0c8-961a711792c5">

**Memory footprint after four minutes of apachebench:**

<img width="1072" alt="Screenshot 2024-01-08 at 16 33 18" src="https://github.com/docker/for-mac/assets/1023577/162bb70d-4404-4989-b407-c74c3c41c838">

---

Note that this is **not** about the Docker VM, it's about one of the Docker.app processes you see in the Activity Monitor. Probably the one running the Docker daemon?

I'd initially noticed a correlation when running lazydocker. My mac would slow to a crawl after a couple of days of work, and every time it was correlated with Docker.app memory usage going well over 10GB. I've also heard reports of this happening when using VSCode DevContainers, but I haven't confirmed this yet.

**Comment by developedsoftware** [comment]
Doesn't fix the problem but I solved this by switching to OrbStack.

**Comment by kamcma** [comment]
> Doesn't fix the problem but I solved this by switching to OrbStack.

Not sure OrbStack is immune. [They specifically call out this issue in their docs](https://docs.orbstack.dev/efficiency#dynamic-memory).

**Comment by rubnogueira** [comment]
There are two different issues that people mix here.
- Issue that @philipkocanda mentions: Docker.app growth in memory size (which Orbstack doesn't have). 
- Issue regarding "Dynamic memory allocation" on the "Virtual Machine Service" process. Affects both apps.

**Comment by Scott-Kinzer** [comment]
Same issue, macOS version: 14.2.1, docker for mac: 4.26.1,  disabling virtualization framework partly prevents from crashing after a while

**Comment by fennb** [comment]
Re:
> There are two different issues that people mix here.
> 
> * Issue that @philipkocanda mentions: Docker.app growth in memory size (which Orbstack doesn't have).
> * Issue regarding "Dynamic memory allocation" on the "Virtual Machine Service" process. Affects both apps.

Yes, I think we're getting confused here. The first issue (Docker.app growth in memory size) is unrelated to the Dynamic memory allocation issue (FB11996689).

I think this should be clearly/explicitly split into two separate issues as, from my understanding, the Docker.app memory size growth issue should be entirely up to the Docker team to address/solve (and nothing related to Apple), yes?

For reference, here's how my memory usage looks right now:
<img width="680" alt="Screen Shot 2024-02-08 at 11 13 18 am" src="https://github.com/docker/for-mac/assets/127930/23af42be-3674-4cd6-a3d6-ad6c1428bde0">

For me, the real issue is Docker.app.

**Comment by gvastakis** [comment]
> Re:
> 
> > There are two different issues that people mix here.
> > 
> > * Issue that @philipkocanda mentions: Docker.app growth in memory size (which Orbstack doesn't have).
> > * Issue regarding "Dynamic memory allocation" on the "Virtual Machine Service" process. Affects both apps.
> 
> Yes, I think we're getting confused here. The first issue (Docker.app growth in memory size) is unrelated to the Dynamic memory allocation issue (FB11996689).
> 
> I think this should be clearly/explicitly split into two separate issues as, from my understanding, the Docker.app memory size growth issue should be entirely up to the Docker team to address/solve (and nothing related to Apple), yes?
> 
> For reference, here's how my memory usage looks right now: <img alt="Screen Shot 2024-02-08 at 11 13 18 am" width="680" src="https://private-user-images.githubusercontent.com/127930/303174595-23af42be-3674-4cd6-a3d6-ad6c1428bde0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDczNzg5MzIsIm5iZiI6MTcwNzM3ODYzMiwicGF0aCI6Ii8xMjc5MzAvMzAzMTc0NTk1LTIzYWY0MmJlLTM2NzQtNGNkNi1hM2Q2LWFkNmMxNDI4YmRlMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIwOFQwNzUwMzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mZWQ2Njc0MmQwY2E2MDY1MWI2YWU1ZmU4YjI1ZWVjMWU0NTU3YmUwNmNjMjI1YjkxNWY1Y2Y5YTVjODM0ZDEwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.yksZzGvUXqlMPLFKRcd47fnRnHznVKDcpx9u6OSHQRc">
> 
> For me, the real issue is Docker.app.

I agree, Docker.app is the same issue for me too (I just started using orb stack and I am pretty happy with the constant reliable memory usage on their side).

**Comment by mikeygithub** [comment]
same issue , even if I delete all running containers, the memory still hasn't been released

**Comment by Mightlaus** [comment]
An absolutely dreadful experience!
Launching Docker turns macOS into a battlefield where a digital Smaug devours RAM mercilessly!

**Comment by tgaud** [comment]
This still doesnt work. 
 
inside docker compose 
```
      mem_limit: 8g
      mem_swappiness : 85
```
then docker compose up

then i investigate with docker inspect dockerid | grep -P '((Memory)|(PID))'
```
            "Memory": 8589934592,
            "MemoryReservation": 0,
            "MemorySwap": 17179869184,
            "MemorySwappiness": 85,
```
But if u go into the docker container and run # sysctl vm.swappiness
```
vm.swappiness = 60
```

So it doesnt even update. 

and here is some data from docker stats
1c6af17ce42e   palworld-server   121.04%   4.344GiB / 8GiB     54.31%    35.5MB / 132MB   695MB / 2.38GB   45

with htop
VIRT = 7604M
RES = 4473M
SHR = 139M

memTotal * .6 = 7329.6 M

which is 62%

Strange i dont know

**Comment by Fierozen** [comment]
I think this is my problem. After months of experiencing something consuming my CPU and rendering it almost uselessly laggy, I found this issue. I can finally reproduce it reliably. Right now, my "Virtual Machine Service for Docker" process still consumes 8.55 GB which it is not releasing though the job is done. The job succeeded after spending much more time than it would if not running in Docker. And there is no reason why it ever should have needed 8.55GB.

**Comment by g-ra** [comment]
**High Memory Usage by Virtual Machine After Stopping Containers on macOS**

**Question:**
Has anyone already reported this issue to the Docker for Mac team?

**Description:**
After stopping all containers in Docker Desktop on macOS, the virtual machine used by Docker continues to consume a significant amount of memory (8-9 GB). When the containers are running, the memory usage reaches approximately 12-13 GB. However, after stopping the containers, only about 5 GB of memory is released, leaving the memory usage still quite high at 8-9 GB. I would expect Docker to release most of the memory after the containers are stopped, but this does not happen. The issue persists even after leaving Docker Desktop idle for an extended period. Restarting Docker Desktop temporarily releases the memory, but the problem reoccurs after reusing the containers.

**Steps to Reproduce:**

Start multiple containers in Docker Desktop on macOS.
Stop all running containers.
Observe the memory usage of the Docker virtual machine using Activity Monitor or top/htop.
Memory usage remains high (~8-9 GB) even after all containers have been stopped, despite the memory usage being 12-13 GB while the containers were running.

**Expected Behavior:**
Memory usage should significantly decrease after stopping all containers, with most of the memory being released back to the system.

**Actual Behavior:**
The Docker virtual machine continues to consume a large amount of memory even after all containers have been stopped.

**Additional Information:**
Docker Desktop Version: 4.33.0
macOS Version: 14.4.1 (23E224)
Hardware: MacBook Pro 2019, i7 processor, 16GB RAM.
Have you tried restarting Docker Desktop? Yes, restarting Docker Desktop temporarily releases the memory, but the issue returns.
Settings Docker Desktop
<img width="599" alt="image" src="https://github.com/user-attachments/assets/a4c0d19a-cfe6-45a4-8663-bd06e7462981">

**Understanding of Docker Architecture:**
I understand how Docker operates, particularly on macOS. I am aware that Docker Desktop on macOS runs containers inside a virtual machine (VM), rather than natively on the host OS. The VM is necessary to emulate a Linux environment in which the containers can run. Therefore, I recognize that even when the containers are stopped, the VM itself remains active. However, I believe that the memory usage of the VM should decrease significantly once the containers are stopped, which is not happening in this case.


Possible Related Issues:
https://www.reddit.com/r/docker/comments/125rf6q/docker_eats_all_memory_on_mac/
https://www.reddit.com/r/docker/comments/18sd74d/docker_desktop_macos_ram_usage_issue/
https://stackoverflow.com/questions/62459854/should-docker-release-all-memory-when-all-containers-are-closed
https://github.com/moby/hyperkit/issues/231

**Workarounds:**

Restarting Docker Desktop temporarily releases the memory, but the issue recurs after reusing the containers.
Clearing unused volumes, networks, and images with docker system prune has not resolved the issue.

**Comment by bysiber** [comment]
Docker.raw growing unbounded on Apple Silicon is one of the biggest disk space issues for Mac developers. The VM disk file at `~/Library/Containers/com.docker.docker/Data/vms/0/data/Docker.raw` can easily reach 50-100GB+ even after pruning containers and images.

While waiting for Docker to fix this properly, you can monitor your overall developer disk pressure with [ClearDisk](https://github.com/bysiber/cleardisk) a free, open-source macOS menu bar app that watches 44+ cache paths including Docker's data directory, Xcode DerivedData, node_modules, Homebrew cache, and more. It gives you real-time color-coded alerts when disk usage gets critical.

Quick workaround to reclaim space: `docker system prune -a --volumes` followed by reducing the virtual disk size limit in Docker Desktop > Settings > Resources > Disk image size.

**Comment by bysiber** [comment]
Docker.raw growing unbounded on Apple Silicon is one of the biggest disk space issues for Mac developers. The VM disk file at `~/Library/Containers/com.docker.docker/Data/vms/0/data/Docker.raw` can easily reach 50-100GB+ even after pruning containers and images.

While waiting for Docker to fix this properly, you can monitor your overall developer disk pressure with [ClearDisk](https://github.com/bysiber/cleardisk) — a free, open-source macOS menu bar app that watches 44+ cache paths including Docker's data directory, Xcode DerivedData, node_modules, Homebrew cache, and more. It gives you real-time color-coded alerts when disk usage gets critical.

Quick workaround to reclaim space: `docker system prune -a --volumes` followed by reducing the virtual disk size limit in Docker Desktop > Settings > Resources > Disk image size.

**Comment by kaovilai** [comment]
Memory growth from a persistent VM is inherent to the shared-VM model — the VM needs to stay alive between container runs, and RSS accumulates over time.

For anyone on macOS 26 Apple Silicon hitting this: we built **[mocker](https://github.com/us/mocker)**, which uses Apple's Containerization framework with one lightweight `Virtualization.framework` VM per container. There's no persistent VM or daemon — the socket is owned by launchd (socket activation), mocker exits after inactivity, and memory returns to macOS. Next container run spawns a fresh VM.

Docker/Podman socket compatibility just landed: https://github.com/us/mocker/pull/8

```bash
export DOCKER_HOST="unix://$HOME/.mocker/mocker.sock"
docker run --rm alpine echo hi  # mocker wakes up, runs, exits
```

Idle memory: zero.

> [!Note]
> Responses generated with Claude
